File: ios/Anemone SDK/general/client/IHttpClient.swift
//
//  IHttpClient.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 5/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public protocol IHttpClient {
    func request(_ httpRequest: HttpRequest) -> Promise<HttpResponse>
}

public extension IHttpClient {
    func get(endpoint: String, params: [String: Any]?) -> Promise<Data> {
        request(HttpRequest.get(url: endpoint, params: params))
            .then { $0.body }
    }

    func post(endpoint: String, params: Any?) -> Promise<Data> {
        request(HttpRequest.post(url: endpoint, params: params))
            .then { $0.body }
    }

    func post(endpoint: String, multipartBodyParams: [MultipartBodyParam], multipartDataFiles: [MultipartDataFile]) -> Promise<Data> {
        request(HttpRequest.post(url: endpoint, multipartBodyParams: multipartBodyParams, multipartDataFiles: multipartDataFiles))
            .then { $0.body }
    }

    func put(endpoint: String, params: Any?) -> Promise<Data> {
        request(HttpRequest.put(url: endpoint, params: params))
            .then { $0.body }
    }
    
    func put(endpoint: String, multipartBodyParams: [MultipartBodyParam], multipartDataFiles: [MultipartDataFile]) -> Promise<Data> {
        request(HttpRequest.put(url: endpoint, multipartBodyParams: multipartBodyParams, multipartDataFiles: multipartDataFiles))
            .then { $0.body }
    }
    
    func patch(endpoint: String, multipartBodyParams: [MultipartBodyParam], multipartDataFiles: [MultipartDataFile]) -> Promise<Data> {
        request(HttpRequest.patch(url: endpoint, multipartBodyParams: multipartBodyParams, multipartDataFiles: multipartDataFiles))
            .then { $0.body }
    }

    func patch(endpoint: String, params: Any?) -> Promise<Data> {
        request(HttpRequest.patch(url: endpoint, params: params))
            .then { $0.body }
    }

    func delete(endpoint: String, params: [String: Any]?) -> Promise<Data> {
        request(HttpRequest.delete(url: endpoint, params: params))
            .then { $0.body }
    }
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/general/client/IHttpClient.kt
package com.anemonesdk.general.client

import com.playtomic.foundation.promise.Promise

/**
 * Created by mgonzalez on 20/12/16.
 */

interface IHttpClient {

    fun request(httpRequest: HttpRequest): Promise<HttpResponse>

    fun get(endpoint: String, params: Map<String, Any>?): Promise<ByteArray> =
        request(HttpRequest.get(url = endpoint, params = params))
            .then(map = { it.body })

    fun post(endpoint: String, params: Any?): Promise<ByteArray> =
        request(HttpRequest.post(url = endpoint, params = params))
            .then(map = { it.body })

    fun post(endpoint: String, multipartBodyParams: List<MultipartBodyParam>, multipartDataFiles: List<MultipartDataFile>): Promise<ByteArray> =
        request(HttpRequest.post(url = endpoint, multipartBodyParams = multipartBodyParams, multipartDataFiles = multipartDataFiles))
            .then(map = { it.body })

    fun put(endpoint: String, params: Any?): Promise<ByteArray> =
        request(HttpRequest.put(url = endpoint, params = params))
            .then(map = { it.body })

    fun put(endpoint: String, multipartBodyParams: List<MultipartBodyParam>, multipartDataFiles: List<MultipartDataFile>): Promise<ByteArray> =
        request(HttpRequest.put(url = endpoint, multipartBodyParams = multipartBodyParams, multipartDataFiles = multipartDataFiles))
            .then(map = { it.body })

    fun patch(endpoint: String, params: Any?): Promise<ByteArray> =
        request(HttpRequest.patch(url = endpoint, params = params))
            .then(map = { it.body })

    fun patch(endpoint: String, multipartBodyParams: List<MultipartBodyParam>, multipartDataFiles: List<MultipartDataFile>): Promise<ByteArray> =
        request(HttpRequest.patch(url = endpoint, multipartBodyParams = multipartBodyParams, multipartDataFiles = multipartDataFiles))
            .then(map = { it.body })

    fun delete(endpoint: String, params: Map<String, Any>?): Promise<ByteArray> =
        request(HttpRequest.delete(url = endpoint, params = params))
            .then(map = { it.body })
}
